### PR TITLE
feat: Google cloud roundtripper for remote write

### PR DESCRIPTION
**Pull request by maxamins**
This PR contains changes related to Google Cloud roundtripper.  It can be used in two modes: 1. Use Default credentials 2. Service account credentials file. [The logic similar to the Azure roudtripper.](https://github.com/prometheus/prometheus/pull/11944) 

Note: Google Cloud doesn't support the RW endpoint yet.

### CODE REVIEW

File: storage/remote/googleiam/README.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+googleiam package
```
Comment (bwplotka) [inline storage/remote/googleiam/README.md:1]: If you want readme for package, probably some main file in the package or `doc.go` would be better? 🤔

File: storage/remote/googleiam/googleiam.go [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+	"context"
+	"fmt"
+	"net/http"
+
+	"golang.org/x/oauth2/google"
+	"google.golang.org/api/option"
+	apihttp "google.golang.org/api/transport/http"
+)
+
+type GoogleIAMConfig struct {
```
Comment (bwplotka) [inline storage/remote/googleiam/googleiam.go:26]: Linter is right we could do `Config` here (and `NewRT` below or something)

**Review by bwplotka — COMMENTED** [review]
Generally looks good, just some linter and readability nits.

Disclaimer: I work with Max at Google Cloud team. We work on making sure GCP supports remote write v2 well, so similar to other cloud providers (azure, aws) we would love to have auth method on remote write. 🤗

File: config/config.go [hunk 3]
```diff
@@ -1105,17 +1107,38 @@ func (c *RemoteWriteConfig) UnmarshalYAML(unmarshal func(interface{}) error) err
 		return err
 	}
 
-	httpClientConfigAuthEnabled := c.HTTPClientConfig.BasicAuth != nil ||
-		c.HTTPClientConfig.Authorization != nil || c.HTTPClientConfig.OAuth2 != nil
-
-	if httpClientConfigAuthEnabled && (c.SigV4Config != nil || c.AzureADConfig != nil) {
-		return fmt.Errorf("at most one of basic_auth, authorization, oauth2, sigv4, & azuread must be configured")
+	err := validateAuthConfigs(c)
+	if err != nil {
```
[the remark this answers is not in the document]
Comment (maxamins) [reply config/config.go]: Good catch :)

File: storage/remote/googleiam/README.md (hunk 1 above)
Comment (maxamins) [reply storage/remote/googleiam/README.md:1]: This is the same as azure https://github.com/prometheus/prometheus/blob/main/storage/remote/azuread/README.md

I followed the same same style for consistency. Let me know if you want me to change it.

File: storage/remote/googleiam/googleiam.go (hunk 2 above)
Comment (maxamins) [reply storage/remote/googleiam/googleiam.go:26]: I tried to keep it consistent with aws and azure. They use the same format https://github.com/prometheus/prometheus/blob/main/storage/remote/azuread/azuread.go#L71-L83. Let me know if you want me to change it.

**Comment by cstyan** [comment]
From the PR name/some comments I assumed this was a generic gcloud IAM roundtripper, but from another comment it sounds like this is specific to google cloud monitoring?

If it's the former, this should live in somewhere more common like the AWS sigv4 code. If the latter, ca we use a clearer name so that it's obvious this isn't reusable for other clients that interact with gcloud.

**Comment by bwplotka** [comment]
That's an amazing point, what do you think @maxamins?

Making it generic IAM client would be probably better (and make potential specifics required against GCM on remote write if there are any?)

File: storage/remote/googleiam/googleiam.go (hunk 2 above)
Comment (bwplotka) [reply storage/remote/googleiam/googleiam.go:26]: I would change it. Just `Config` and `NewRT`

**Comment by maxamins** [comment]
This can serve as a generic gcloud IAM roundtripper. I considered adding it to common, but was following the precedent set by the Azure round tipper (see [comment](https://github.com/prometheus/prometheus/pull/11944#issuecomment-1431377550)) which says it's kept here since it's currently unused elsewhere unlike the AWS round tipper. 
 
Can we merge this change for now and move it to common if future use cases arise? What do you think?

File: storage/remote/googleiam/README.md (hunk 1 above)
Comment (bwplotka) [reply storage/remote/googleiam/README.md:1]: Yea, it's weird. Weird formatting and only one sentence. I would just kill this file and put that in a single Go file we have as a package comment (on `googleiam.go`) https://go.dev/blog/godoc#:~:text=Comments%20on%20package%20declarations%20should%20provide%20general%20package%20documentation

**Comment by bwplotka** [comment]
Happy with that 👍🏽 @cstyan ?

**Comment by cstyan** [comment]
> set by the Azure round tipper (see https://github.com/prometheus/prometheus/pull/11944#issuecomment-1431377550) which says it's kept here since it's currently unused elsewhere unlike the AWS round tipper.

:+1: fine by me

**Comment by cstyan** [comment]
@roidelapluie any chance you can take a look here? the test failures are for:
```
goleak: Errors on successful test run: found unexpected goroutines:
[Goroutine 16 in state select, with go.opencensus.io/stats/view.(*worker).start on top of the stack:
```
in packages that already have the `testutil.TolerantVerifyLeak(m)` call added

**Comment by bwplotka** [comment]
Rebase might help if something was fixed recently

**Comment by cstyan** [comment]
> Rebase might help if something was fixed recently

Not recent afaict, something from ~2020 that is supposed to be ignoring goroutine leaks from dependencies but it's not working here with Max's changes.

**Comment by maxamins** [comment]
So the error was related to lingering goroutine is related to OpenCensus's statistics collection worker in the `"google.golang.org/api/transport/http` package. This package is used in other locations and the errors are tolerated via ` testutil.TolerantVerifyLeak(m)`. I did the same to the promql, rules and tsdb and that resolved the error.

**Comment by cstyan** [comment]
> So the error was related to lingering goroutine is related to OpenCensus's statistics collection worker in the `google.golang.org/api/transport/http` package. This package is used in other locations and the errors are tolerated via `testutil.TolerantVerifyLeak(m)`. I did the same to the promql, rules and tsdb and that resolved the error.

Not sure how I missed this when we were talking on slack the other day :+1: nice.

**Review by cstyan — APPROVED** [review]

**Review by bwplotka — APPROVED** [review]

**Pull request merged**